### DISCUSSION TITLE
YoloV5 model accuracy issue

### DISCUSSION THREAD

**Discussion by Sreekanth-aa** [ultralytics/yolov5#5676] [Q&A]
Hi,.
We are using Yolov5 for object detection. We have 294 images and 4 classes, when we try to train Yolov5l with batch size of >20 and Yolov5x with batch size of >10 the models are not able to train as the process is getting killed. The models are trained in Google colab with RAM size of 12Gb. We would like to know what should be the max RAM size required for Yolov5 model for training if we need to increase the batch size.

We were able to train the Yolov5m with custom dataset (294 images and 4 classes) on colab, when we test the model with --conf 0.1 the model was not able to predict few class images, but with --conf 0.05 the model was able to predict the object but the prediction was wrong. We would like to know how we can increase the accuracy of the model.

Thanks,
Sreekanth

**Comment by glenn-jocher** [comment]
@Sreekanth-aa 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

[Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/](https://github.com/ultralytics/yolov5/issues/2844#issuecomment-851338384)

**Reply by Sreekanth-aa** [reply]
Hi Glenn,

Thanks for getting back to us.

We would like to know the RAM size required for training the Yolov5x6 model. While training the model in colab we are getting the below error, can you please let us know why we are getting this error.

`RuntimeError: CUDA out of memory. Tried to allocate 26.00 MiB (GPU 0; 11.17 GiB total capacity; 10.36 GiB already allocated; 3.81 MiB free; 10.57 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF`

**Reply by glenn-jocher** [reply]
@Sreekanth-aa 👋 Hello! Thanks for asking about **CUDA memory issues**. YOLOv5 🚀 can be trained on CPU, [single-GPU](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data), or [multi-GPU](https://github.com/ultralytics/yolov5/issues/475). When training on GPU it is important to keep your batch-size small enough that you do not use all of your GPU memory, otherwise you will see a CUDA Out Of Memory (OOM) Error and your training will crash. You can observe your CUDA memory utilization using either the `nvidia-smi` command or by viewing your console output:

<img width="800" alt="Screenshot 2021-05-28 at 12 19 51" src="https://user-images.githubusercontent.com/26833433/131122318-8141d1ce-13de-44bf-bbdc-487bad442ff0.png">

### CUDA Out of Memory Solutions

If you encounter a CUDA OOM error, the steps you can take to reduce your memory usage are:

- Reduce `--batch-size`
- Reduce `--img-size`
- Reduce model size, i.e. from YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s
- Train with [multi-GPU](https://github.com/ultralytics/yolov5/issues/475) at the same `--batch-size`
- Upgrade your hardware to a larger GPU
- Train on free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

### AutoBatch

You can use [YOLOv5 AutoBatch](https://github.com/ultralytics/yolov5/pull/5092) (NEW) to find the best batch size for your training by passing `--batch-size -1`. AutoBatch will solve for a 90% CUDA memory-utilization batch-size given your training settings. AutoBatch is experimental, and only works for Single-GPU training. It may not work on all systems, and is not recommended for production use.

<img width="1746" alt="Screenshot 2021-11-06 at 12 31 10" src="https://user-images.githubusercontent.com/26833433/140608327-9c0c0baf-a0ee-4551-a4ff-5ee26b155c51.png">

Good luck and let us know if you have any other questions!
